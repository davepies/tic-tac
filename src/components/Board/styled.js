import styled from 'styled-components';

export const BoardRow = styled.div`
    display: flex;
`;

export const Status = styled.div`
    margin-bottom: 10px;
`;

export const ResetButton = styled.button`
    background: green;
`;