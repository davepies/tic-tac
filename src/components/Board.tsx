import React from 'react';
import styled from 'styled-components';

import { calculateWinner } from '../helpers';

import Square from './Square';

const BoardRow = styled.div`
    clear: both;
    content: "";
    display: table;
`;

const Status = styled.div`
    margin-bottom: 10px;
`;

const ResetButton = styled.button`
    background: green;
`;

interface IBoardState {
    squares: (string | null)[];
    xIsNext: boolean
}

class Board extends React.Component<{}, IBoardState> {
    initialState: IBoardState = {
        squares: [
            null, null, null,
            null, null, null,
            null, null, null
        ],
        xIsNext: true
    }

    state: IBoardState = { ...this.initialState };

    resetGame() {
        this.setState(this.initialState)
    }

    updateGame(squareToUpdateIndex: number): void {
        const { squares, xIsNext } = this.state;
        const nextSquares = [...squares];


        // square has already been set
        if (nextSquares[squareToUpdateIndex] != null) {
            return;
        }

        nextSquares[squareToUpdateIndex] = xIsNext ? 'X' : 'O';

        this.setState({
            squares: nextSquares,
            xIsNext: !xIsNext
        });
    }

    renderSquare(i: number): JSX.Element {
        return (
            <Square
                value={this.state.squares[i]}
                onClick={() => this.updateGame(i)}
            />
        );
    }

    renderGameOver(winner: string): JSX.Element {
        return (
            <div>
                We have a winner!!! {winner} earns 1.0000000 AUD
                    <ResetButton onClick={() => this.resetGame()}>Play again</ResetButton>
            </div>
        )
    }

    render() {
        const { squares, xIsNext } = this.state;
        const winner = calculateWinner(squares);

        if (winner) {
            return this.renderGameOver(winner);
        }

        const statusMessage = `Next player: ${xIsNext ? 'X' : 'O'}`;

        return (
            <>
                <Status>{statusMessage}</Status>
                <BoardRow>
                    {this.renderSquare(0)}
                    {this.renderSquare(1)}
                    {this.renderSquare(2)}
                </BoardRow>
                <BoardRow>
                    {this.renderSquare(3)}
                    {this.renderSquare(4)}
                    {this.renderSquare(5)}
                </BoardRow>
                <BoardRow>
                    {this.renderSquare(6)}
                    {this.renderSquare(7)}
                    {this.renderSquare(8)}
                </BoardRow>
            </>
        );
    }
}

export default Board;