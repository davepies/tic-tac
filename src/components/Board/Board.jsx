import React from 'react';

import { calculateWinner } from '../../common/helpers';

import Square from '../Square';
import { BoardRow, Status, ResetButton } from './styled';

class Board extends React.Component {

    initialState = {
        squares: [
            null, null, null,
            null, null, null,
            null, null, null
        ],
        xIsNext: true,
        history: [[
            null, null, null,
            null, null, null,
            null, null, null
        ]],
        historyIndex: 0
    }

    state = { ...this.initialState };

    resetGame() {
        this.setState(this.initialState)
    }

    updateGame(squareToUpdateIndex) {
        const { squares, xIsNext, history, historyIndex } = this.state;
        const nextSquares = [...squares];
        const nextHistory = [...history];
        const nextHistoryIndex = historyIndex + 1;

        // the square has already been set
        if (nextSquares[squareToUpdateIndex] != null) {
            return;
        }

        nextSquares[squareToUpdateIndex] = xIsNext ? 'X' : 'O';
        nextHistory.push(nextSquares);

        this.setState({
            history: nextHistory,
            historyIndex: nextHistoryIndex,
            squares: nextSquares,
            xIsNext: !xIsNext
        });
    }

    backInHistory() {
        const { history, historyIndex } = this.state;

        const nextHistoryIndex = (historyIndex - 1) > -1 ? historyIndex - 1 : historyIndex;
        const nextSquares = history[nextHistoryIndex];

        this.setState({
            historyIndex: nextHistoryIndex,
            squares: nextSquares
        });
    }

    renderSquare(i) {
        return (
            <Square
                value={this.state.squares[i]}
                onClick={() => this.updateGame(i)}
            />
        );
    }

    renderGameOver(winner) {
        return (
            <div>
                We have a winner!!! {winner} earns 1.0000000 AUD
                    <ResetButton onClick={() => this.resetGame()}>Play again</ResetButton>
            </div>
        )
    }

    render() {
        const { squares, xIsNext } = this.state;
        const winner = calculateWinner(squares);

        if (winner) {
            return this.renderGameOver(winner);
        }

        const statusMessage = `Next player: ${xIsNext ? 'X' : 'O'}`;

        return (
            <>
                <Status>{statusMessage}</Status>
                <BoardRow>
                    {this.renderSquare(0)}
                    {this.renderSquare(1)}
                    {this.renderSquare(2)}
                </BoardRow>
                <BoardRow>
                    {this.renderSquare(3)}
                    {this.renderSquare(4)}
                    {this.renderSquare(5)}
                </BoardRow>
                <BoardRow>
                    {this.renderSquare(6)}
                    {this.renderSquare(7)}
                    {this.renderSquare(8)}
                </BoardRow>
                <button onClick={() => this.backInHistory()}>Back in History</button>
            </>
        );
    }
}

export default Board;