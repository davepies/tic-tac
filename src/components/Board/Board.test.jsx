import React from 'react';
import { shallow } from 'enzyme';

import Board from './';

test('<Board /> renders', () => {
    const wrapper = shallow(<Board />);
    expect(wrapper).toMatchSnapshot();
});

test('<Board /> contains 3 rows', () => {
    const wrapper = shallow(<Board />);
    expect(wrapper.find('styled__BoardRow').length).toEqual(3);
});

test('<Board /> contains status message', () => {
    const wrapper = shallow(<Board />);
    expect(wrapper.find('styled__Status').text().length).toBeGreaterThan(0);
});

test('<Board /> has player "X" starging', () => {
    const wrapper = shallow(<Board />);
    expect(wrapper.find('styled__Status').text()).toMatch('Next player: X');
});