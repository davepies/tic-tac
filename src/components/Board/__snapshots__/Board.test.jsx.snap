// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Board /> renders 1`] = `
<Fragment>
  <styled__Status>
    Next player: X
  </styled__Status>
  <styled__BoardRow>
    <Square
      onClick={[Function]}
      value={null}
    />
    <Square
      onClick={[Function]}
      value={null}
    />
    <Square
      onClick={[Function]}
      value={null}
    />
  </styled__BoardRow>
  <styled__BoardRow>
    <Square
      onClick={[Function]}
      value={null}
    />
    <Square
      onClick={[Function]}
      value={null}
    />
    <Square
      onClick={[Function]}
      value={null}
    />
  </styled__BoardRow>
  <styled__BoardRow>
    <Square
      onClick={[Function]}
      value={null}
    />
    <Square
      onClick={[Function]}
      value={null}
    />
    <Square
      onClick={[Function]}
      value={null}
    />
  </styled__BoardRow>
  <button
    onClick={[Function]}
  >
    Back in History
  </button>
</Fragment>
`;
